clippy_lints/src/single_field_patterns.rs

```diff
@@ -0,0 +1,264 @@
+                SingleField::Unused => {
+                    spans.push((target, String::from("_")));
+                },
+                SingleField::Id { pattern, .. } => {
+                    if let Some(str) = snippet_opt(cx, pattern) {
+                        spans.push((target, str));
+                    } else {
+                        return None;
+                    }
+                    if let Some(one) = the_one {
+                        if sf != one {
+                            return None;
+                        }
+                    } else {
+                        the_one = Some(sf);
+                    }
+                },
+            }
+        } else {
+            return None;
+        }
+    }
+    the_one.map(|one| (one, spans))
+}
+
+fn apply_lint_sf(cx: &LateContext<'_>, span: Span, sugg: impl IntoIterator<Item = (Span, String)>) {
+    span_lint_and_then(
+        cx,
+        SINGLE_FIELD_PATTERNS,
+        span,
+        "this single-variant pattern only matches one field",
+        |diag| {
+            multispan_sugg_with_applicability(diag, "try this", Applicability::MachineApplicable, sugg);
+        },
+    );
+}
+
+fn remove_deref<'a>(mut scrutinee: &'a Expr<'a>) -> &'a Expr<'a> {
+    // it would be wrong to convert something like `if let (x, _) = *a` into `if let x = *a.0`
+    while let ExprKind::Unary(UnOp::Deref, expr) = scrutinee.kind {
+        scrutinee = expr;
+    }
+    scrutinee
+}
+
+fn lint_sf<'hir>(
+    cx: &LateContext<'_>,
+    overall_span: Span,
+    scrutinee: &Expr<'_>,
+    patterns: impl Iterator<Item = &'hir Pat<'hir>>,
+) {
+    if scrutinee.span.from_expansion() {
+        return;
+    }
+    let scrutinee_name = if let Some(name) = snippet_opt(cx, remove_deref(scrutinee).span) {
+        name
+    } else {
+        return;
+    };
+    match cx.typeck_results().expr_ty(scrutinee).kind() {
+        ty::TyKind::Adt(def, ..) if def.is_struct() => {
+            if let Some((field, mut spans)) = find_sf_lint(cx, patterns, &struct_sf) {
+                spans.push((
+                    scrutinee.span,
+                    match field {
+                        SingleField::Id { id, .. } => format!("{}.{}", scrutinee_name, id.as_str()),
+                        SingleField::Unused => return,
+                    },
+                ));
+                apply_lint_sf(cx, overall_span, spans);
+            }
+        },
+        _ => (),
+    };
+}
+
+impl LateLintPass<'_> for SingleFieldPatterns {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &Expr<'tcx>) {
+        if expr.span.from_expansion() {
+            return;
+        }
+        match IfLetOrMatch::parse(cx, expr) {
+            Some(IfLetOrMatch::Match(scrutinee, arms, MatchSource::Normal)) => {
+                lint_sf(cx, expr.span, scrutinee, arms.iter().map(|arm| arm.pat));
+            },
+            Some(IfLetOrMatch::IfLet(scrutinee, pat, ..)) => lint_sf(cx, expr.span, scrutinee, once(pat)),
```

Comment:
Big NIT, but it took me some time to figure out what the `once` function does, until I remembered that it is a function from `iter`.

```suggestion
            Some(IfLetOrMatch::IfLet(scrutinee, pat, ..)) => lint_sf(cx, expr.span, scrutinee, iter::once(pat)),
```